docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,253 @@
+
+While the interfaces of the P2P layer and most of the node package are
+already internalized, this precludes several kinds of operational
+patterns that are important to key tendermint users who use tendermint
+as a library, specifically introspecting the tendermint node service
+and replacing components is not supported in the latest version of the
+code, and some of these use cases would require maintaining a vendor
+copy of the code. Adding these features requires rather extensive
+(internal/implementation) changes to the `node` and `rpc` packages,
+and this ADR describes a model for changing the way that tendermint
+nodes initialize, in service of providing this kind of functionality.
+
+We consider node initialization, because the current implemention
+provides strong connections between all components, as well as between
+the components of the node and the RPC layer, and being able to think
+about the interactions of these components will help enable these
+features and help define the requirements of the node package.
+
+## Alternative Approaches
+
+### Do Nothing
+
+The current implementation is functional and sufficient for the vast
+majority of use cases (e.g. all users of the Cosmos-SDK as well as
+anyone who runs tendermint and the ABCI application in separate
+processes.) In the current implementation, and even previous versions,
+modifying node initialization or injecting custom components required
+copying most of the `node` package, which amounts to requiring users
+to maintain a vendored copy of tendermint for these kinds of use
+cases.
+
+While this is (likely) not tenable in the long term, as users do want
+more modularity, and the current service implementation is brittle and
+difficult to maintain, it may be viable in the short and medium term
+to delay implementation which will allow us to do more product
+research and build more confidence and consensus around the eventual
+solution.
+
+### Generic Service Plugability
+
+We can imagine a system design that exports interfaces (in the Golang
+sense) for all components of the system, to permit runtime dependency
+injection of all components in the system so that users can compose
+tendermint nodes of arbitrary user supplied components.
+
+This is an interesting proposal, and may be interesting to persue, but
+essentially requires doing a lot more API design and increases the
+surface area of our API. While this is not a goal at the moment,
+eventually providing support for some kinds of plugability may be
+useful, so the current solution does not explicitly forclose the
+possibility of this alternative.
+
+### Abstract Dependency Based Startup and Shutdown
+
+The proposal on in this document simplifies and makes tendermint node
+initialization more abstract, but the system lacks a number of
+features which daemon/service initialization might provide, such as a
+dependency based system that allows the authors of services to control
```

Comment:
I'm not totally sure I know what a dependency based system means, mind clarifying?

Reply:
I think, as an alternative, you could imagine a system where when you add a service to the node, you specify identifiers of services that it depends on (for initialization) and you could use this information to control the order of service initialization (and also potentially add parallelism.) 

I think it's a bit overwrought for our use case, but it would be "the upgrade pick" if that makes sense.